### PR TITLE
refactor to support init options and context

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/rmw#154

### CODE REVIEW

File: rmw_fastrtps_cpp/src/rmw_init.cpp [hunk 1]
```diff
@@ -12,13 +12,88 @@
...
+rmw_ret_t
+rmw_shutdown(rmw_context_t * context)
 {
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eprosima_fastrtps_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context->impl, RMW_RET_INVALID_ARGUMENT);
```
Comment (sloretz) [inline rmw_fastrtps_cpp/src/rmw_init.cpp]: Same comment about `context->impl` should be `null`

File: rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp [hunk 2]
```diff
@@ -12,13 +12,88 @@
...
+rmw_init_options_init(rmw_init_options_t * init_options, rcutils_allocator_t allocator)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(init_options, RMW_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ALLOCATOR(&allocator, return RMW_RET_INVALID_ARGUMENT);
+  if (NULL != init_options->implementation_identifier) {
+    RMW_SET_ERROR_MSG("expected zero-initialized init_options");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  init_options->instance_id = 0;
+  init_options->implementation_identifier = eprosima_fastrtps_identifier;
```
Comment (sloretz) [inline rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp]: Should there be an `eprosima_fastrtps_dynamic_identifier` like connext has?

File: rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp [hunk 3]
```diff
@@ -12,13 +12,88 @@
...
+rmw_ret_t
+rmw_shutdown(rmw_context_t * context)
 {
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eprosima_fastrtps_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context->impl, RMW_RET_INVALID_ARGUMENT);
```
Comment (sloretz) [inline rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp]: same comment about `context->impl` being set to `null` in `rmw_init()`

File: rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp (hunk 2 above)
Comment (wjwwood) [reply rmw_fastrtps_dynamic_cpp/src/rmw_init.cpp]: Maybe, but I think that would be follow up to the fastrtps dynamic pr's and not in scope for this pr.

**Pull request merged**